### ISSUE TITLE
issue with new reduce_dimension

### ISSUE THREAD

**Issue by clausmichele** [Open-EO/openeo-processes-dask#35]
Here I did a demo of doing a local processing based on openeo-processes-dask + the new graph parser. https://github.com/clausmichele/openeo-python-client/blob/client_side_proc/examples/notebooks/Client_Side_Processing/test_local_collections.ipynb
I was using the versions `openeo_pg_parser_networkx==2022.11.0 openeo_processes_dask==2022.12.4` and it worked fine.
Now I am trying to update the workflow with the latest changes and I found an issue with the `reduce_dimension` step.
I attach the code to reproduce the issue and the required files:
[test_data_reduce_dimension.zip](https://github.com/Open-EO/openeo-processes-dask/files/10457880/test_data_reduce_dimension.zip)
```python
import xarray as xr
import inspect
import logging
import importlib
import json
import openeo_pg_parser_networkx
from openeo_pg_parser_networkx.graph import OpenEOProcessGraph
from openeo_pg_parser_networkx import ProcessRegistry
import openeo_processes_dask
from openeo_processes_dask.core import process

print(openeo_pg_parser_networkx.__version__)
print(openeo_processes_dask.__version__)

2022.12.1
2023.1.0
```

```python
standard_processes = [
    func
    for _, func in inspect.getmembers(
        importlib.import_module("openeo_processes_dask.process_implementations"),
        inspect.isfunction,
    )
]
process_registry = ProcessRegistry(wrap_funcs=[process])
for func in standard_processes:
    process_registry[func.__name__] = func

#We need to define a custom `load_collection` process, used to load local netCDFs
logger = logging.getLogger(__name__)
def load_local_collection(*args, **kwargs):
    pretty_args = {k: type(v) for k, v in kwargs.items()}
    logger.warning(f"Running process load_collection")
    logger.warning(f"kwargs: {pretty_args}")
    logger.warning("-" * 80)    
    data = xr.open_dataset(kwargs['id'],decode_coords='all').to_array(dim='bands')
    return data
process_registry["load_collection"] = load_local_collection

with open('vv_vh_ratio.json') as f:
    pg = json.load(f)

ratio_openeo_client = OpenEOProcessGraph(pg).to_callable(process_registry)()
```

output:

```python
Deserialised process graph into nested structure
Running process load_local_collection
kwargs: {'parameters': <class 'dict'>, 'id': <class 'str'>, 'spatial_extent': <class 'NoneType'>, 'temporal_extent': <class 'NoneType'>}
--------------------------------------------------------------------------------
Running process load_collection
kwargs: {'parameters': <class 'dict'>, 'id': <class 'str'>, 'spatial_extent': <class 'NoneType'>, 'temporal_extent': <class 'NoneType'>}
--------------------------------------------------------------------------------

Walking node root-0607f718-e29a-4c07-bcb8-d660d5fe9e1f
Walking node log1-f2be5e5b-f3c6-4b07-a44e-433a07cd14be
Walking node divide1-f2be5e5b-f3c6-4b07-a44e-433a07cd14be
Walking node arrayelement1-f2be5e5b-f3c6-4b07-a44e-433a07cd14be
Walking node arrayelement2-f2be5e5b-f3c6-4b07-a44e-433a07cd14be
Walking node reducedimension1-0607f718-e29a-4c07-bcb8-d660d5fe9e1f
Walking node mean1-22336cca-dc17-43f2-93c6-e21c741744ee
Walking node loadcollection1-0607f718-e29a-4c07-bcb8-d660d5fe9e1f

Running process reduce_dimension
kwargs: {'parameters': <class 'dict'>, 'data': <class 'xarray.core.dataarray.DataArray'>, 'dimension': <class 'str'>, 'reducer': <class 'functools.partial'>}
--------------------------------------------------------------------------------

---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
File ~\Miniconda3\envs\sap\lib\site-packages\openeo_pg_parser_networkx\graph.py:353, in OpenEOProcessGraph._map_node_to_callable.<locals>.node_callable(parent_callables, **kwargs)
    351 try:
    352     # If this node has already been computed once, just grab that result from the results_cache instead of recomputing it.
--> 353     return results_cache.__getitem__(node)
    354 except KeyError:

KeyError: 'reducedimension1-0607f718-e29a-4c07-bcb8-d660d5fe9e1f'

During handling of the above exception, another exception occurred:

TypeError                                 Traceback (most recent call last)
Input In [22], in <cell line: 1>()
----> 1 ratio_openeo_client = OpenEOProcessGraph(pg).to_callable(process_registry)()

File ~\Miniconda3\envs\sap\lib\site-packages\openeo_pg_parser_networkx\graph.py:349, in OpenEOProcessGraph._map_node_to_callable.<locals>.node_callable(parent_callables, **kwargs)
    346 def node_callable(parent_callables, **kwargs):
    347     # The node needs to first call all its parents, so that results are prepopulated in the results_cache
    348     for func in parent_callables:
--> 349         func(**kwargs)
    351     try:
    352         # If this node has already been computed once, just grab that result from the results_cache instead of recomputing it.
    353         return results_cache.__getitem__(node)

File ~\Miniconda3\envs\sap\lib\site-packages\openeo_pg_parser_networkx\graph.py:372, in OpenEOProcessGraph._map_node_to_callable.<locals>.node_callable(parent_callables, **kwargs)
    369 if not dynamic_parameters:
    370     dynamic_parameters = kwargs
--> 372 result = prebaked_process_impl(**dynamic_parameters)
    374 results_cache[node] = result
    376 return result

File ~\Miniconda3\envs\sap\lib\site-packages\openeo_processes_dask\core.py:80, in process.<locals>.wrapper(positional_parameters, named_parameters, *args, **kwargs)
     77 logger.warning(f"kwargs: {pretty_args}")
     78 logger.warning("-" * 80)
---> 80 return f(*resolved_args, **resolved_kwargs)

File ~\Miniconda3\envs\sap\lib\site-packages\openeo_processes_dask\process_implementations\cubes\reduce.py:23, in reduce_dimension(data, reducer, dimension, context, **kwargs)
     18 if dimension not in data.dims:
     19     raise DimensionNotAvailable(
     20         f"Provided dimension ({dimension}) not found in data.dims: {data.dims}"
     21     )
---> 23 reduced_data = data.reduce(reducer, dim=dimension, keep_attrs=True, context=context)
     25 # Preset
     26 if "reduced_dimensions_min_values" not in data.attrs:

File ~\Miniconda3\envs\sap\lib\site-packages\xarray\core\dataarray.py:3658, in DataArray.reduce(self, func, dim, axis, keep_attrs, keepdims, **kwargs)
   3614 def reduce(
   3615     self: T_DataArray,
   3616     func: Callable[..., Any],
   (...)
   3622     **kwargs: Any,
   3623 ) -> T_DataArray:
   3624     """Reduce this array by applying `func` along some dimension(s).
   3625 
   3626     Parameters
   (...)
   3655         summarized data and the indicated dimension(s) removed.
   3656     """
-> 3658     var = self.variable.reduce(func, dim, axis, keep_attrs, keepdims, **kwargs)
   3659     return self._replace_maybe_drop_dims(var)

File ~\Miniconda3\envs\sap\lib\site-packages\xarray\core\variable.py:1947, in Variable.reduce(self, func, dim, axis, keep_attrs, keepdims, **kwargs)
   1943     if isinstance(axis, tuple) and len(axis) == 1:
   1944         # unpack axis for the benefit of functions
   1945         # like np.argmin which can't handle tuple arguments
   1946         axis = axis[0]
-> 1947     data = func(self.data, axis=axis, **kwargs)
   1948 else:
   1949     data = func(self.data, **kwargs)

TypeError: node_callable() got multiple values for argument 'parent_callables'
```

**Comment by LukeWeidenwalker** [comment]
Ah - I haven't had time to replicate this yet, but looking at it I have a suspicion that this is something that can occur with subprocess graphs that have multiple nodes. The fact that it worked before might have been a happy accident, rather than it breaking in the latest release. I'll investigate as soon as I can and report back here!

**Comment by LukeWeidenwalker** [comment]
I've just confirmed my suspicions - there are two problems that contribute to this;
1) `args` get swallowed by `node_callable` in the pg parser, so the `data` args that gets created by `.reduce` and all these gets lost on the way to the process implementation. I have a fix for this ready to go, will open a PR in the parser today.
2) parameters don't get handed all the way through subgraphs. This is a known implementation gap that I haven't had time to address yet - since it's coming up both for you and @ValentinaHutter in UC6 related work, I'll devote the time to implement this asap - keep you posted!

**Comment by LukeWeidenwalker** [comment]
Thanks for reporting - this should be fixed in the latest release (2023.1.1)! I've also added the process graph to the integration tests to make sure this doesn't break again! :)

**Comment by clausmichele** [comment]
I confirm that now the provided code works fine again.
